### PR TITLE
C implementation of STM32RTC class

### PR DESCRIPTION
**Pull request by MrZloHex**
Hello,
I was writing a firmware for STM32 in C and I encountered a problem that there are no C libraries for internal RTC,
and i tried to rewrite your STM32RTC class from C++ into C Struct. In the near future i will add more functionality to be close enough to your implementation.
I tested it on STM32G070RB - and it's all okay.

With respect, Alexey.

### CODE REVIEW

**Comment by MrZloHex** [comment]
- [X] Import all methods

**Comment by fpistm** [comment]
Hi @MrZloHex 
Is your firmware is to be used in Arduino ecosytem?

As a first look your porting change the default import of the Arduino library: STM32RTC.h becomes STM32RTC.hpp so the default examples can't compile as no C++ API are available.

**Comment by MrZloHex** [comment]
Yes, I'm using arduino ecosystem.

I made STM32RTC.h into .hpp to differentiate C andd C++ implementation.
I could change C impl into STM32RTC_c.h and make C++ version as STM32RTC.h if you want.
Or simply change examples.

**Comment by fpistm** [comment]
> Or simply change examples.

No, several examples are available outside this library and several users/fw using this library include only the STM32RTC.h file. So this is not an options.

Honestly I wonder if it is a good idea to mix/duplicate code in this library. It is really confusing. Just talking about how Arduino include the library, it uses the STM32RTC.h as main entry to include it. So how you would manage this to only include the C part ? 
Maybe the only way would be to have only one.h file and using `#ifdef __cplusplus`

**Comment by MrZloHex** [comment]
I merged 2 headers into 1 with `__cplusplus` and STB-style macro to divide C++ and C implementations.
Now all examples work and to use C version before include you should `#define C_IMPLEMENTATION` to confirm that you to use c version.

File: src/STM32RTC.h [hunk 1]
```diff
@@ -34,6 +34,9 @@
   ******************************************************************************
   */
 
+#if defined(__cplusplus) && !defined(C_IMPLEMENTATION)
```
Comment (fpistm) [inline src/STM32RTC.h:37]: I don't think it is needed to add an extra definition
```suggestion
#if defined(__cplusplus)
```

File: src/STM32RTC.h [hunk 2]
```diff
@@ -258,3 +261,959 @@ class STM32RTC {
...
+
+typedef unsigned char  u8;
+typedef unsigned short u16;
+typedef unsigned int   u32;
+typedef long unsigned int ul32;
+typedef unsigned long  u64;
+typedef signed char    i8;
+typedef signed short   i16;
+typedef signed int     i32;
+typedef signed long    i64;
```
Comment (fpistm) [inline src/STM32RTC.h:298]: Avoid to define new type. Use standard one: uint8_t, int16_t,....
Please replace them in all definition and declaration

**Review by fpistm — CHANGES_REQUESTED** [review]
An example using C interface could be fine to add.

**Comment by fpistm** [comment]
Closed as no update since a while.

**Pull request closed without merging**